toolz/functoolz.py

```diff
@@ -681,3 +659,172 @@ def __name__(self):
             return '%s_excepting_%s' % (self.f.__name__, exc_name)
         except AttributeError:
             return 'excepting'
+
+
+def is_valid_args(func, args, kwargs):
```

Comment:
Could we lift the `if` checks to import time to reduce the impact of using curry. something like:


```python
if PY3:
    def is_valid_args(f, args, kwargs):
        try:
            sig = inspect.signature(func)
        except ValueError:
            return _is_builtin_valid_args(func, args, kwargs)
        except TypeError:
            return False
        try:
            sig.bind(*args, **kwargs)
        except TypeError:
            return False
        return True

    if PY34 or PYPY:
        def is_valid_args(f, args, kwargs, _is_valid_args=is_valid_args):
            # Python 3.4 or pypymay lie so use our registry for
            # builtins instead
            val = _is_builtin_valid_args(func, args, kwargs)
            if val is not None:
                return val

            return _is_valid_args(f, args, kwargs)

else:
    def is_valid_args(f, args, kwargs):
        try:
            spec = inspect.getargspec(func)
        except TypeError:
            return _is_builtin_valid_args(func, args, kwargs)
        defaults = spec.defaults or ()
        num_pos = len(spec.args) - len(defaults)
        missing_pos = spec.args[len(args):num_pos]
        if any(arg not in kwargs for arg in missing_pos):
            return False

        if spec.varargs is None:
            num_extra_pos = max(0, len(args) - num_pos)
        else:
            num_extra_pos = 0

        kwargs = dict(kwargs)

        # Add missing keyword arguments (unless already included in `args`)
        missing_kwargs = spec.args[num_pos + num_extra_pos:]
        kwargs.update(zip(missing_kwargs, defaults[num_extra_pos:]))

        # Convert call to use positional arguments
        args = args + tuple(kwargs.pop(key) for key in spec.args[len(args):])

        return not (
            not spec.keywords and kwargs or
            not spec.varargs and len(args) > len(spec.args) or
            set(spec.args[:len(args)]) & set(kwargs)
        )

    if PYPY:
        def is_valid_args(f, args, kwargs, _is_valid_args=is_valid_args):
            # pypy may lie so use our registry for builtins instead
            val = _is_builtin_valid_args(func, args, kwargs)
            if val is not None:
                return val

            return _is_valid_args(f, args, kwargs)


is_valid_args.__doc__ = """\
Is ``func(*args, **kwargs)`` a valid function call?

This function relies on introspection and does not call the function.
Returns None if validity can't be determined.

>>> def add(x, y):
...     return x + y

>>> is_valid_args(add, (1,), {})
False
>>> is_valid_args(add, (1, 2), {})
True
>>> is_valid_args(map, (), {})
False

**Implementation notes**
Python 2 relies on ``inspect.getargspec``, which only works for
user-defined functions.  Python 3 uses ``inspect.signature``, which
works for many more types of callables.

Many builtins in the standard library are also supported.
"""
```